core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1576,6 +1582,32 @@ private Optional<Throwable> acknowledgeCompleteBatch(
         return Optional.empty();
     }
 
+    // The caller of this function is expected to hold lock.writeLock() when calling this method.
+    protected void updateEndOffsetAndResetFetchOffsetMetadata(long updatedEndOffset) {
+        endOffset = updatedEndOffset;
+        fetchOffsetMetadata = Optional.empty();
+    }
+
+    protected void updateFetchOffsetMetadata(Optional<LogOffsetMetadata> fetchOffsetMetadata) {
+        lock.writeLock().lock();
+        try {
+            this.fetchOffsetMetadata = fetchOffsetMetadata;
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    protected Optional<LogOffsetMetadata> fetchOffsetMetadata() {
+        lock.readLock().lock();
+        try {
+            if (findNextFetchOffset.get())
+                return Optional.empty();
```

Comment:
This means `tryComplete` will never get non-empty `fetchOffsetMetadata` and its calculation of minBytes will be off. We need to think through how to address this.